### ISSUE TITLE
Define release process

### ISSUE THREAD

**Issue by embee8** [homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform#94]
The package is currently published to npm using my access token (scoped to the `homebridge-panasonic-ac-platform` package). Given my changing role in the project, I'd like to add additional maintainers to the npm package, and for them to update the GitHub secret so that I can phase out the old access token. @jandersonhill, do you have an npm account I could add?

**Comment by jandersonhill** [comment]
I'd not used npm until last night, but have just created an account :-)

On the topic of npm, is there an easy way to amend the npm workflow to give the option of releasing a beta? Short term, I guess I could modify the workflow to `- run: npm publish --tag beta`, but wondered if there was a better way.

**Comment by embee8** [comment]
I might have some notes on this which I will refine and pack into a development guide attached to the repository.

**Comment by jandersonhill** [comment]
Thanks, @embee8 :-)

**Comment by jandersonhill** [comment]
BTW If you could help me to create a new token, that would be fantastic!

**Comment by embee8** [comment]
See https://docs.npmjs.com/creating-and-viewing-access-tokens and let me know if that helps.

**Comment by embee8** [comment]
Also, please hold off from publishing new versions for a short while. I see some things have been tangled up, e.g. the version of `master` not matching the published package on npm anymore. I will clean things up.

**Comment by jandersonhill** [comment]
Could that be because I'm publishing versions of the code to npm from branches of the master that are still in beta? It could equally be because I screwed something up, it seems I have a bit of a learning curve!

**Comment by embee8** [comment]
Every time you create a new release on GitHub, the publishing workflow runs. When that happens, npm will verify the version number it finds in package.json, and you'll see that it hasn't been changed from 1.6.1, and you cannot release a version twice as it would introduce inconsistencies.

I'll try to make this as easy to handle as possible for you. Would you prefer to manage releases and tags through the command line or the GitHub UI?

**Comment by jandersonhill** [comment]
Seriously, much appreciated. I think I'm better off sticking with the GitHub UI for now... I updated the version number in package.json for the two beta branches, but I've left the master alone.

**Comment by embee8** [comment]
Bear with me - I'll figure something out and come back for feedback.

**Comment by mkz212** [comment]
@embee8 @jandersonhill 

I am not a specialist in this but I have such an idea. Let there be two branches - master and beta. If a normal edition (not a beta) is to be published, let it take from a master. When it is to be beta, the source should be a beta branch.

**Comment by embee8** [comment]
Take a look at [CONTRIBUTING.md](https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/dev-guidelines/CONTRIBUTING.md) and let me know what you think.

**Comment by mkz212** [comment]
@embee8 As for me, that's how every tutorial should look like! 👍

**Comment by mkz212** [comment]
@embee8 

Forgive me if any question seems silly. I have never published a plugin and I would like to know what it looks like.

- version number is to enter by user who make PR? And how does he know what number to enter? For example, there is currently version 1.2.3, is it supposed to introduce 1.2.4 or 1.3.0? Etc.
- you recommend that there be one branch: master. If you make a merge to it, and then want to pull back some merge, can you do it in some easy way?
- if you publish something to npm, can you delete it later? (E.g. if something by mistake)

The rest seems simple. And a great solution with this two publish workflows !👍

**Comment by embee8** [comment]
**Versioning**
We are applying semantic versioning, see [here](https://semver.org/). Backward compatible bug fixes receive a patch version bump (1.3.2 -> 1.3.3). Backward compatible new features receive a minor version bump (1.3.2 -> 1.4.0). Changes that break backward compatibility receive a major version bump. This is also documented in the contribution guide.

Because the version number has semantic meaning, the decision lies with the author of the changes but can be challenged/corrected by the reviewers who eventually merge the changes.

**Tags**
Every PR has a package.json in a way. If the PR doesn't make changes to it, the base branch version will be used. Workflow checks on PRs will point out to increase the version.

**Merging and publishing**
There are limited way to unpublish packages from npm (see [here](https://docs.npmjs.com/unpublishing-packages-from-the-registry)), but this shouldn't be a routine practice to make up for a lack of testing or a robust release process.

Beta releases are here for people to test the latest features. They don't replace testing for developers on their local installations before creating a pull request.

One cannot "pull back a merge" in the sense of simply undoing it. But one can undo the changes by submitting a new PR and merge those changes again, which will go into a new version. This is clean and simple, in contrast to a sophisticated branching model that I don't think would serve this project well.

**Comment by mkz212** [comment]
@embee8 Thanks! Very understandable 👍

I just can't understand one thing. When we do the merge and we have the version set in package.json, how do we create this tag from the version (which is later selected in the workflow as a source)? In some other project I see in Releases page that there are also tags and you can remove them but there is no option to add a tag.

**Comment by embee8** [comment]
On GitHub, tags can be created in the process of creating releases. I'd like to automate this because it's a fiddly job which is easily messed up. By handling the creation of tags and releases in a workflow, maintainers don't have to do it manually.

If you don't introduce any breaking changes, the next version should be 1.8.0.

**Comment by mkz212** [comment]
@embee8 I also think it would be nice if it was a little more automated 👍

**Comment by embee8** [comment]
@jandersonhill, @mkz212, given the new release process, we need to consolidate the two feature branches ([1.7.1](https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/tree/1.7.1) and [1.8.0-beta](https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/tree/1.8.0-beta)) into `master`.

Could you please merge the feature branches with the master branch in your repos, correct any linter errors, and submit a PR to the upstream repository?

Once all that is done, we'll delete the feature branches and proceed as discussed above.

**Comment by mkz212** [comment]
@embee8  Can you merge them? And then I will update my fork?

**Comment by jandersonhill** [comment]
Thanks for putting this together, @mkz212. This process makes sense.

I may have some questions later today, but I have my hands full with two energetic boys at the moment!

**Comment by mkz212** [comment]
@Embee8 did the greatest job preparing these two workflows for publication and beta publication! I envy you. For Christmas, I will ask a movie about how this workflows works in practice! ;)

**Comment by embee8** [comment]
> @embee8  Can you merge them? And then I will update my fork?

Let me see what I can do. 👍

**Comment by mkz212** [comment]
@embee8 I will be very grateful! Thanks for the help!

**Comment by mkz212** [comment]
@embee8 @jandersonhill Set everything and it is time for the first beta version 1.8.0 and user tests! I can't wait! 👍😀

**Comment by jandersonhill** [comment]
@embee8 Some quick thoughts/questions:

For clarity:

- We want to merge 1.7.1 and 1.8.0-beta into the master before doing anything else (however, see my question below);
- Each future PR will have its own branch for testing, to be merged into the master when deemed stable;
- Every (accepted) PR will result in a new version number at the point of release (either as beta or production).

Assumptions:

- Public beta (as opposed to local) testing requires that a new version number is tied to a specific PR
- The version number along with the PR's description should be documented in release notes;
- Public testing of each PR should done incrementally as opposed to concurrently (otherwise this could easily get confusing).

Questions:

- Should we merge 1.7.1 into the master and retain the 1.8.0 branch for a public beta (given neither of us have a way to test these new features)?
- Are betas automatically released without the 'latest release' flag on npm, so as to avoid people auto-upgrading?

If we're agreed on what to do with 1.8.0 (whether to merge this into the master), I can do this, although the release may have happen at some point tomorrow in order to follow the new workflow. I agree we should automate as much as possible for future releases.

Sorry if any of these questions sound stupid!

**Comment by mkz212** [comment]
@jandersonhill 

Beta are always released without latest flag. If someone has a stable version, he will not see the beta version notification. He has to click install another version and there he will see only the beta. If someone already on the installed beta, they will see information about the update (for both beta and stable).

@embee8 @jandersonhill 

I don't want to go too deep into it but... Definitely a great solution with these workflows for publication. But I think it shouldn't be that every new PR is a new version. Sometimes as part of working on a version you have to add a few PR.

**Comment by mkz212** [comment]
@embee8 Release process test 👍

![IMG_4473](https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/assets/82271669/053a6df6-2fec-4b58-9978-841e81a659d6)

**Comment by embee8** [comment]
Almost there – just cleaning up a few things now to cover all angles.

**Comment by mkz212** [comment]
@embee8 @jandersonhill 

Testing 1.8.0 first beta. 

Oscilate swing:
nanoe - ok!
swing - ok!

Value on start
nanoe - ok!
swing - ok!

I dont have eco navi and inside cleaning. But even when i set turn on eco navi and inside cleaning there is no errors! 

👍👍👍

**Comment by mkz212** [comment]
@jandersonhill Can you set up this pinned Issue for beta testers? And invite to test?

**Comment by embee8** [comment]
Perfect! Though, please open another Issue for v1.8.0 testing if necessary. Let's keep this one focused on the release process.

**Comment by mkz212** [comment]
@embee8 

Maybe we will continue discussions about the release process here but in a closed Issue?

We are planning with @jandersonhill to create a pinned Issue with a description of new features and information for testers. At the same time, the closure of all existing Issues because either the problem has already been solved, or the expired or the functions just introduced concerned this (and here invite the participants to the tests).

**Comment by jandersonhill** [comment]
I'm happy either way - either to treat 1.8.x as a release that streamlines the workflow and go to a new version number for the new feature, or to combine. I can understand both arguments, although changing the version number might be cleaner.

**Comment by embee8** [comment]
@jandersonhill, do you have any remaining questions about the release process? Let me know if the documentation is clear enough.

**Comment by jandersonhill** [comment]
@embee8, my main concerns are getting to grips with how npm works outside the Github environment and whether the new workflow using the Github CLI can be replicated within the UI until I'm comfortable with using the CLI. What I don't want to do is screw this up!

**Comment by embee8** [comment]
The good news is that you won't have to use the npm CLI to manage releases, see [here](https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/blob/master/CONTRIBUTING.md#releasing-a-new-version).

You only have to use the Git commands outlined in the docs if you want to make changes in the code, not to merge pull requests or release new versions. GitHub allows to edit files too, but it's always best to make changes locally, lint, compile, and test your changes before submitting a pull request.

For clarity: Git and GitHub are two different things – one is a version control system, the other a platform offering (Git) version control and other services around it.

**Comment by jandersonhill** [comment]
Ok, I can manage that!

Your help transitioning the project is greatly appreciated :-) I'm also aware we need to look at the npm token next...

**Comment by mkz212** [comment]
@embee8 @jandersonhill Will any of you update Readme? Should I do PR?

**Comment by jandersonhill** [comment]
@mkz212 Just catching up, could you do a PR and I'll deal with this?
@embee8 We should look at the options for the npm account

I also agree, public threads probably needn't contain the minutia of all behind the scenes activities and we should have a private thread for some of these conversations

**Comment by mkz212** [comment]
@jandersonhill You can't do PR to Issue. Just copy paste and add something from yourself if you want.


New beta released! 1.8.0. New functions. Support for Eco Navi, Nanoe etc. 

New functions:
- Choice of what to control Oscilate Switch: swing, nanoe, eco navi, inside cleaning.
- Choice what value should be at each AC start for: swing, nanoe, eco navi, inside cleaning.
- When the internal unit does not have a temperature sensor, the temperature of the external unit is display, and when there are no both, it is automatically depending on the mode (heating / cooling).

What do we need from beta testers?
- What of these functions does your AC have: swing, nanoe, eco navi, inside cleaning?
- Does your AC have a temperature sensor in the internal and external unit? How does it work for you?
- Enable Debug in plugin settings and in the Homebridge settings and paste the response from your AC (it will be in logs) and write what is the status of a given function (swing, nanoe, eco navi, inside cleaning) for this response.
- Do the introduced new functions work for you?

**Comment by jandersonhill** [comment]
@mkz212 I was responding to your question: 'Will any of you update Readme? Should I do PR?'

**Comment by embee8** [comment]
Regarding the old issues: I recommend checking each of them individually and leaving a comment like "This should be fixed in version x.x.x. Please install the latest version and report back if the issue persists".
***
Regarding gathering 1.8.0 beta feedback: @mkz212, please feel free to create the issue yourself as developer of 1.8.0, since you have the most context and will be able to incorporate any feedback.
***
> @embee8 @jandersonhill Will any of you update Readme? Should I do PR?

You mean regarding the new features in 1.8.0? This should have been part of the initial PR already. If it wasn‘t, please submit it as follow-up PR. When someone writes new features, the expectation is for then to document their own changes.
***
> P.S.: I think we can also close this post and continue in a closed thread.

The point of closing an issue is to mark the conversation as closed. If there‘s more to discuss, keep it open or leave it open until the job is done or the fix is in production.

**Comment by embee8** [comment]
> @embee8 We should look at the options for the npm account

Is [this](https://www.npmjs.com/~jandersonhill) your npm profile? I'll add you as a collaborator of the package and then you'll be able to create an access token which you can set in the repository settings on GitHub.

**Comment by jandersonhill** [comment]
@embee8 Yes, that's the one I created a few days ago

**Comment by jandersonhill** [comment]
> > @embee8 We should look at the options for the npm account
> 
> Is [this](https://www.npmjs.com/~jandersonhill) your npm profile? I'll add you as a collaborator of the package and then you'll be able to create an access token which you can set in the repository settings on GitHub.

@embee8 Yes, that's the one I created a few days ago

**Comment by jandersonhill** [comment]
> Regarding the old issues: I recommend checking each of them individually and leaving a comment like "This should be fixed in version x.x.x. Please install the latest version and report back if the issue persists".
> 
> Regarding gathering 1.8.0 beta feedback: @mkz212, please feel free to create the issue yourself as developer of 1.8.0, since you have the most context and will be able to incorporate any feedback.
> 
> > @embee8 @jandersonhill Will any of you update Readme? Should I do PR?
> 
> You mean regarding the new features in 1.8.0? This should have been part of the initial PR already. If it wasn‘t, please submit it as follow-up PR. When someone writes new features, the expectation is for then to document their own changes.


In an attempt to make this bit a little easier, I've edited @mkz212 proposed comments for the readme as follows  - we can change the feature to be tested so we do one at a time (the wording in italics is taken from the current readme.md so we can see where it would fit within the existing doc):

_...
- If the plugin affects the general responsiveness and reliability of your Homebridge setup, you can run it as an isolated [child bridge](https://github.com/homebridge/homebridge/wiki/Child-Bridges)._

## Latest Beta

Beta versions of this plugin are made available from time to time for those who wish to test new features. Please be aware that there may be issues we've yet to discover that need addressing (see 'Contributing' below for reporting any issues you find). The current beta release version with its changes is listed below. To install the latest beta on Homebridge, click on the three vertical dots to the right of 'Settings' for your current Homebridge Panaonic AC Platform and choose 'Install Alternate Version'.

New beta released! 1.8.0. New functions include support for a lack of external temperature sensor

New functions:

- ...

To help with development, we would be grateful to get feedback from beta testers on the following:

- ...
- Are the new functions working as described?

_## Contributing
..._

**Comment by jandersonhill** [comment]
I would propose the first feature to be tested could be:

_When the internal unit does not have a temperature sensor, the temperature of the external unit is display, and when there are no both, it is automatically depending on the mode (heating / cooling)._

To clarify, wasn't the issue being addressed a lack of a sensor in the **outdoor** unit? Also, are there indoor units with no temperature sensors? I'm not sure how they would work without one.

**Comment by embee8** [comment]
I'll leave this to you, but I'd personally keep the README.md general, and put information about specific releases into a separate file, or simply pack it into a GitHub issue which is even easier to manage.

Regarding npm package ownership: You should be good to go.

**Comment by jandersonhill** [comment]
Agreed, I wouldn't want the readme to become a change log, but I feel referencing (any) current beta version might be helpful if it's somewhere visible. I definitely think it's worth putting this in the GitHub issue and will update accordingly :-)

**Comment by jandersonhill** [comment]
@mkz212 Can I check you're ok with my wording about the beta above (and also whether you meant a lack of indoor temperature sensor, or whether that should read a lack of an outdoor sensor)?

Are you ok with us asking beta testers to test one new feature at a time?

**Comment by mkz212** [comment]
@jandersonhill  I don't mind. As for the temperature sensor, it's about the sensor in the internal unit - so they are without - see #31

**Comment by embee8** [comment]
Closing out this issue as we clarified the release process. I recommend moving the beta 1.8.0 conversation into another thread.

@jandersonhill, no rush, but you can replace `NPM_TOKEN` in the [repository settings](https://github.com/homebridge-panasonic-ac-platform/homebridge-panasonic-ac-platform/settings/secrets/actions) with your own token.

FYI: `GH_CLI_TOKEN` will expire on 19 Nov 2024 at which point the release workflow will start to fail and the token needs to be replaced. You can create a new one [here](https://github.com/organizations/homebridge-panasonic-ac-platform/settings/personal-access-tokens/active).
